### PR TITLE
test(bom): harden BOM version-freshness gate (#1576)

### LINKED ISSUE #1576: Harden BOM version-freshness gate (bidirectional parse, docs-only PRs, type-safety)
## Summary

Follow-up hardening for the BOM version-freshness gate introduced in PR #1572 (which resolved #1424). The gate's core path works — a registry version bump that forgets `make bom-docs` now fails CI — but three edge cases remain where it can false-pass or misbehave. None block #1572; tracking them here.

## F1 — Freshness test is one-way and duplicate-unsafe

`TestCommittedBOMVersionsMatchRegistry` (`tools/bom/freshness_test.go`) only checks registry → document. It never compares the reverse direction, so:

- **Stale/removed rows go undetected.** Removing a component from `recipes/registry.yaml` while leaving its row in `docs/user/container-images.md` still passes.
- **Duplicate rows, last-wins.** `parseBOMVersionTable` walks every pipe table matching the `Component`/`Pinned Version` headers and does `out[name] = ver`, so a later duplicate row silently overwrites the generated value. It parses the whole doc, not just the generated markers.

Fix: parse only the generated BOM section (between its markers), reject duplicate component rows, and compare the registry and document component sets bidirectionally before comparing versions.

## F2 — Freshness test never runs on docs-only PRs

`.github/workflows/merge-gate.yaml` classifies a PR as `code=false` when every changed file is `**.md`/`docs/**`/`LICENSE`, routing it to `tests-skip`. A PR that edits only `docs/user/container-images.md` therefore never executes the freshness test, despite `docs/contributor/recipe.md` stating version freshness is gated. The intended registry-bump path is covered (it touches `registry.yaml` → `code=true`), so this is the narrower docs-only hand-edit hole.

Fix: add a lightweight path-filtered job that runs the freshness test when `docs/user/container-images.md` or `recipes/registry.yaml` change, or soften the doc's "gated" wording to match reality.

## F4 — Not deployment-type-safe

- **(a) Latent:** the BOM generator (`tools/bom/main.go`) emits `Version` only from `Helm.DefaultVersion` and never from `Kustomize.DefaultTag`, but the freshness test's expected value falls back to `Kustomize.DefaultTag`. The first Kustomize component carrying a `defaultTag` would fail the freshness test even after `make bom-docs`. No such component exists today, so it cannot trigger yet.
- **(b) Hardening:** `assertExemptionDefaultsInstalled` (`pkg/recipe/version_pin_guard_test.go`) accepts `ref.Version == def || ref.Tag == def`. For a Helm component it should compare only the active field; an inactive `Tag` matching `def` could spoof the sole-consumer proof.

Fix: select the active version field by effective component type and reject the incompatible field, in both the generator/test expectation and the exemption proof.

## F3 — Example recipes are guarded by neither gate

The hand-maintained example recipes in `examples/recipes/*.yaml` bypass overlay resolution, and neither new guard examines them: `TestOverlayVersionPinsMatchRegistry` scans only `store.Base`/`store.Overlays`/`store.Mixins`, and `TestCommittedBOMVersionsMatchRegistry` compares only the registry against `docs/user/container-images.md`. PR #1572 refreshed the examples by hand (nvsentinel, gpu-operator, AKS kube-prometheus-stack pins, and a stale `driver.version` override), so they are consistent today — but a future registry or overlay bump can leave the examples advertising a stale chart version while both gates stay green, reintroducing the exact drift class #1424 targets, on the examples surface.

Fix: cover `examples/recipes/*` with a guard or a regeneration/consistency check (e.g. resolve each example against the registry and fail on undeclared divergence), or explicitly document them as an unguarded illustrative snapshot boundary if that is the intended contract.

## F5 — Base is not considered a standalone consumer in the sole-consumer proof

`assertExemptionDefaultsInstalled` (`pkg/recipe/version_pin_guard_test.go`) resolves only overlays with non-nil `Spec.Criteria`; `base.yaml` pins kube-prometheus-stack to the `84.4.0` registry default but carries no criteria, so it is excluded from the resolved `results`. The AKS exemption therefore stays satisfied only because a criteria-bearing non-AKS overlay inherits base's default. This holds today (and the test fails loudly if no overlay carries criteria), but the sole-consumer guarantee is coupled to overlay structure rather than stated explicitly.

Fix: either count `store.Base` as a standalone consumer in the proof, or add a code comment documenting why base is intentionally excluded and what invariant keeps the exemption honest. Related to F4(b) (active-field selection in the same proof).

## Context

- Introduced by #1572 / resolves the follow-up dimension of #1424.
- All three are non-blocking edge cases; the gate's realistic threat model (registry bump without regen) is enforced.

### PR DESCRIPTION
**Pull request by yuanchen8911**
## Summary

Follow-up hardening for the BOM version-freshness gate added in #1572. Closes three edge cases in the new gate where it could false-pass or misbehave: the committed-BOM check is now an exact bidirectional projection, it runs on docs-only PRs, and version selection is deployment-type-safe.

## Motivation / Context

#1572 satisfied #1424 for the realistic threat (a registry bump that forgets `make bom-docs` fails CI). A cross-review of the merged gate surfaced three residual edge cases, tracked in #1576, none of which blocked #1572 but all of which weaken the gate:

- **F1** — `TestCommittedBOMVersionsMatchRegistry` checked only registry → doc and scanned every pipe table in the file, so a stale row for a removed component passed, and a duplicate row could shadow the generated value.
- **F2** — the merge gate classifies Markdown/docs-only changes as non-code and skips the full `tests` job, so a PR editing only `docs/user/container-images.md` never ran the freshness check despite the docs claiming freshness is gated.
- **F4** — the BOM generator emitted a pinned version only from Helm fields while the freshness test's expectation fell back to `Kustomize.DefaultTag` (so the first Kustomize component would fail after `make bom-docs`), and the exemption sole-consumer proof accepted either `Version` or `Tag` (an inactive field could spoof a default installation).

Fixes: #1576
Related: #1424, #1572

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [x] Build/CI/tooling

## Component(s) Affected

- [x] Recipe engine / data (`pkg/recipe`)
- [x] Docs/examples (`docs/`, `examples/`)
- [x] Other: BOM generator + tests (`tools/bom`), merge-gate workflow (`.github/workflows`)

## Implementation Notes

- **F1** (`tools/bom/freshness_test.go`) — parse only the generated `<\!-- BEGIN/END AICR-BOM -->` marker section; reject duplicate component rows; compare the registry and doc component sets **bidirectionally** (missing row *and* orphan row both fail) before comparing versions. Adds `freshness_parse_test.go` unit tests for marker extraction and the table parser.
- **F2** (`.github/workflows/merge-gate.yaml`) — add a path-filtered `bom-freshness` job (+ skip companion, wired into the aggregate gate) that runs `TestCommittedBOMVersionsMatchRegistry` when `docs/user/container-images.md`, `recipes/registry.yaml`, or the generator sources change, so the gate holds on docs-only PRs.
- **F4** (`tools/bom/main.go`, `pkg/recipe/version_pin_guard_test.go`) — the generator emits the pinned version by effective component type (Helm `defaultVersion` or Kustomize `defaultTag`); the exemption proof selects the active field by type instead of accepting either.
- Docs: `docs/contributor/recipe.md` updated to describe the bidirectional check and the docs-only gate coverage.

## Testing

```bash
go test -race ./tools/bom/... ./pkg/recipe/...   # ok
golangci-lint run -c .golangci.yaml ./tools/bom/... ./pkg/recipe/...   # 0 issues
actionlint .github/workflows/merge-gate.yaml     # OK
yamllint .github/workflows/merge-gate.yaml       # OK
```

`TestCommittedBOMVersionsMatchRegistry` verifies 25 pinned components bidirectionally against the committed BOM; `TestOverlayVersionPinsMatchRegistry` verifies 38 pins (one declared exemption). Draft pending full `make qualify`.

## Risk Assessment

- [x] **Low** — Test/CI/generator hardening only; no change to recipe resolution or bundle output. The generator change is a no-op for the current registry (no Kustomize component carries a `defaultTag`), and the committed BOM regenerates with zero diff.

**Rollout notes:** N/A — no runtime behavior change.

## Checklist

- [x] Tests pass locally (`make test` with `-race`)
- [x] Linter passes (`make lint`)
- [x] I did not skip/disable tests to make CI green
- [x] I added/updated tests for new functionality
- [x] I updated docs if user-facing behavior changed
- [x] Changes follow existing patterns in the codebase
- [x] Commits are cryptographically signed (`git commit -S`)

### CODE REVIEW

File: tools/bom/freshness_parse_test.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	t.Run("returns empty for a section with no components table", func(t *testing.T) {
+		got, err := parseBOMVersionTable("just prose, no table\n")
+		if err != nil {
+			t.Fatalf("unexpected err: %v", err)
+		}
+		if len(got) != 0 {
+			t.Errorf("got %v, want empty", got)
+		}
+	})
+}
```
[the remark this answers is not in the document]
Comment (yuanchen8911) [reply tools/bom/freshness_parse_test.go:132]: Done in 78d6b188 — TestParseBOMVersionTable is now a single {name, section, want, wantErr} table with one t.Run loop, matching TestExtractGeneratedSection.

File: tools/bom/freshness_test.go [hunk 2]
```diff
@@ -93,13 +151,28 @@ func TestCommittedBOMVersionsMatchRegistry(t *testing.T) {
...
+	if begin < 0 {
+		return "", fmt.Errorf("missing %q marker", bomBeginMarker)
+	}
+	begin += len(bomBeginMarker)
+	end := strings.Index(doc[begin:], bomEndMarker)
+	if end < 0 {
+		return "", fmt.Errorf("missing %q marker after %q", bomEndMarker, bomBeginMarker)
+	}
+	return doc[begin : begin+end], nil
+}
```
[the remark this answers is not in the document]
Comment (yuanchen8911) [reply tools/bom/freshness_test.go:188]: Leaving these as fmt.Errorf. They are local errors inside a _test.go helper that are consumed immediately by t.Fatalf, never propagated to production callers, so they carry no error code. pkg/errors.New requires a code argument (errors.New(errors.ErrCodeInvalidRequest, msg)); the suggested errors.New(fmt.Sprintf(...)) would not compile. Adding a synthetic code to a test-only sentinel adds noise without value, so fmt.Errorf is the right fit here.

File: tools/bom/main.go [hunk 3]
```diff
@@ -197,15 +215,16 @@ func renderHelmComponent(ctx context.Context, repoRoot string, c component, r he
...
 		DisplayName: c.DisplayName,
 		Type:        c.kind(),
 		Repository:  c.Helm.DefaultRepository,
 		Chart:       c.Helm.DefaultChart,
-		Version:     c.Helm.DefaultVersion,
+		Version:     version,
 		Namespace:   c.Helm.DefaultNamespace,
-		Pinned:      c.Helm.DefaultVersion != "",
+		Pinned:      version != "",
 	}
```
[the remark this answers is not in the document]
Comment (yuanchen8911) [reply tools/bom/main.go:238]: Fixed in 78d6b188 — strict mode now flags an unpinned Kustomize component too: the check is (r.Type == kindHelm || r.Type == kindKustomize) && r.Version == "". Since pinnedVersion is now type-aware, an empty defaultTag yields Version=="" and is caught the same way an unpinned Helm chart is.

**Comment by yuanchen8911** [comment]
Follow-up (out of scope for this PR): resolver rejects ComponentRefs with incoherent type/field combinations.

Several review findings on this PR trace to one root cause: nothing validates that a resolved `ComponentRef` has a coherent type/field combination — e.g. `Type: Helm` carrying a Kustomize `Tag`/`Path`, a Kustomize ref without a `Path`, or a `Tag` without a source. This leaks into the signed attestation BOM (`pkg/evidence/attestation/bom.go` `BuildAutoBOM` records the declared-type metadata even when the deployers would build the other type) and is why this PR carries defensive, deploy-shape-aware checks in the exemption proof (`pkg/recipe/version_pin_guard_test.go`) and type-routed CycloneDX metadata (`pkg/bom/bom.go`).

The proper fix is resolver-level validation: enforce coherence once in `finalizeRecipeResult` (`pkg/recipe/metadata_store.go`, right after `ApplyRegistryDefaults`), mirroring the rules already in `pkg/bundler/deployer/localformat/writer.go` (Kustomize requires `Path`; `Tag` requires a `Repository`) plus the inverse (a Helm ref must not carry `Tag`/`Path`). That fixes the attestation issue at the source and would let the test-side deploy-shape mirroring added here be simplified later.

It is a separate, self-contained change (a small `validateComponentRefCoherence` helper + one call site + tests) deliberately kept out of this PR because it runs on every recipe resolution — its blast radius warrants its own focused review and CI run. Note: it is entirely latent today (no Kustomize component with a tag/path exists in the registry or any recipe).

Tracking here until a dedicated issue is filed.

**Comment by yuanchen8911** [comment]
Follow-up now tracked as #1584 (Bug, P2) — resolver-level validation of ComponentRef type/field coherence.

File: .github/workflows/merge-gate.yaml [hunk 4]
```diff
@@ -319,6 +332,52 @@ jobs:
...
+          fi
+
+  bom-freshness-skip:
+    needs: [check-paths]
+    if: needs.check-paths.outputs.bom != 'true'
+    runs-on: ubuntu-latest
+    timeout-minutes: 1
+    steps:
+      - run: echo "No BOM/registry changes — BOM freshness check not required"
+
```
[the remark this answers is not in the document]
Comment (yuanchen8911) [reply .github/workflows/merge-gate.yaml:380]: Declining: no job in this workflow sets a job-level name: — tests, analyze, malware-scan, actionlint, and the verify-renovate/verify-renovate-skip pair are all anonymous. zizmor flags every job equally, not just these two. Adding name: to only bom-freshness/bom-freshness-skip would break the workflow's consistent convention; if we want named jobs it should be a separate sweep across all of them.

File: pkg/bom/bom.go [hunk 5]
```diff
@@ -152,7 +152,15 @@ func BuildBOM(meta Metadata, results []ComponentResult) *cdx.BOM {
...
+			// Name the version property by component type so a Kustomize
+			// component's Git tag is not mislabeled as a Helm chart version
+			// (the CycloneDX doc already declares aicr:component:type). Helm
+			// and manifest components keep aicr:helm:version.
+			versionProp := "aicr:helm:version"
+			if r.Type == "kustomize" {
+				versionProp = "aicr:kustomize:tag"
+			}
+			props = append(props, cdx.Property{Name: versionProp, Value: r.Version})
 		}
```
[the remark this answers is not in the document]
Comment (yuanchen8911) [reply pkg/bom/bom.go:182]: Done in 422d7c01 — exported TypeHelm/TypeKustomize/TypeManifest from pkg/bom; tools/bom's kindHelm/kindKustomize/kindManifest now alias them and component.kind() returns those constants, so the string values can't drift across the two packages. (Note the renderer also compares with strings.EqualFold, so it already tolerates the recipe ComponentType enum's capitalized "Kustomize" from the attestation builder, which a shared literal alone wouldn't cover.)

File: tools/bom/main.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (yuanchen8911) [reply tools/bom/main.go:238]: Correction to my earlier reply: this change was reverted in a later push — strict mode is Helm-only again (main.go:116, `r.Type == kindHelm && r.Version == ""`). Expanding `-strict` to Kustomize conflicts with ADR-006, which scopes strict chart-version pinning to Helm, and `ComponentRegistry.Validate()` accepts a tagless Kustomize component — so flagging it here would fail `make lint` (bom-pinning-check runs `-strict`) on a valid entry with a misleading "chart version is not pinned." Left as-is by design.

**Review by mchmarny — APPROVED** [review]
